Lib/typing.py

```diff
@@ -1057,6 +1114,219 @@ def __dir__(self):
+        # == Too many args ==
+
+        typevars=(T1,),     args=(int, str)           => TypeError
+        typevars=(T1,),     args=(*tuple[int, str],)  => TypeError
+        # We could have two ints, so this might be too many.
+        typevars=(T1,),     args=(*tuple[int, ...],)  => TypeError
+
+        # ===== Two TypeVars =====
+
+        typevars=(T1, T2),  args=(int, str)           => {T1: int, T2: str}
+        typevars=(T1, T2),  args=(*tuple[int, str],)  => {T1: int, T2: str}
+
+        # ===== A single TypeVarTuple =====
+
+        typevars=(Ts,),     args=()                   => {Ts: ()}
+        typevars=(Ts,),     args=(int,)               => {Ts: (int,)}
+        typevars=(Ts,),     args=(int, str)           => {Ts: (int, str)}
+        typevars=(Ts,),     args=(*tuple[()],)        => {Ts: ()}
+        typevars=(Ts,),     args=(*tuple[int],)       => {Ts: (int,)}
+        typevars=(Ts,),     args=(*tuple[int, str],)  => {Ts: (int, str)}
+        typevars=(Ts,),     args=(*tuple[int, ...],)  => {Ts: (*tuple[int, ...])}
+
+        # ===== A single TypeVar and a single TypeVarTuple =====
+
+        # == Too few args ==
+
+        typevars=(T, Ts),   args=()                  => TypeError
+        typevars=(T, Ts),   args=(*tuple[()],)       => TypeError
+        # Again, this means ">= 0 ints", but we need ">= 1".
+        typevars=(T, Ts),   args=(*tuple[int, ...])  => TypeError
+
+        # == Right number of args ==
+
+        typevars=(T, Ts),   args=(int,)              => {T: int, Ts: ()}
+        typevars=(T, Ts)    args=(int, str)          => {T, int, Ts: (str,)}
+        typevars=(T, Ts),   args=(*tuple[int])       => {T: int, Ts: ()}
+        typevars=(T, Ts),   args=(*tuple[int, str])  => {T: int, Ts: (str,)}
+        typevars=(T, Ts),   args=(int, *tuple[str, ...])
+                                            => {T: int, Ts: (*tuple[str, ...],)}
+        typevars=(T, Ts),   args=(*tuple[int], *tuple[str, ...])
+                                            => {T: int, Ts: (*tuple[str, ...],)}
+    """
+
+    # This would be a pretty complicated algorithm in its full generality.
+    # Fortunately, we can make two observations, which simplify things:
+    # 1. As of PEP 646, there can only be a maximum of one TypeVarTuple.
+    # 2. When considering unpacked arbitrary-length tuples
+    #    like *tuple[int, ...], we can't assign them or any part of them to
+    #    regular TypeVars:
+    #    * We can't assign the whole *tuple[int, ...] to a single TypeVar
+    #      because...well, it represents an arbitrary number of types, and a
+    #      TypeVar only holds exactly one type.
+    #    * We can't assign one of the `int`s to a TypeVar, because we can't
+    #      be sure there'll be any `int`s at all: tuple[int, ...] is a tuple
+    #      of *zero* or more ints.
+
+    if not typevars:
+        return {}
+
+    num_typevartuples = len([x for x in typevars
+                             if isinstance(x, TypeVarTuple)])
```

Comment:
```suggestion
    num_typevartuples = sum(1 for x in typevars
                            if isinstance(x, TypeVarTuple))
```
Avoid intermediate list

Reply:
Done.